src/core/border_router/routing_manager.cpp

```diff
@@ -3110,6 +3161,172 @@ void RoutingManager::RsSender::HandleTimer(void)
     return;
 }
 
+#if OPENTHREAD_CONFIG_BORDER_ROUTING_DHCP6_PD_ENABLE
+RoutingManager::PdPrefixManager::PdPrefixManager(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mEnabled(false)
+    , mTimer(aInstance)
+{
+    mPrefix.Clear();
+}
+
+Error RoutingManager::PdPrefixManager::GetPrefixInfo(PrefixTableEntry &aInfo) const
+{
+    Error     error = kErrorNone;
+    TimeMilli now   = TimerMilli::GetNow();
+
+    VerifyOrExit(IsRunning() && HasPrefix(), error = kErrorNotFound);
+
+    aInfo.mPrefix              = mPrefix.GetPrefix();
+    aInfo.mValidLifetime       = mPrefix.GetValidLifetime();
+    aInfo.mPreferredLifetime   = mPrefix.GetPreferredLifetime();
+    aInfo.mMsecSinceLastUpdate = now - mPrefix.GetLastUpdateTime();
+
+exit:
+    return error;
+}
+
+void RoutingManager::PdPrefixManager::WithdrawPrefix(void)
+{
+    VerifyOrExit(HasPrefix());
+
+    LogInfo("Withdrew platform provided outdated prefix: %s", mPrefix.GetPrefix().ToString().AsCString());
+
+    mPrefix.Clear();
+    mTimer.Stop();
+
+    Get<RoutingManager>().ScheduleRoutingPolicyEvaluation(kImmediately);
+
+exit:
+    return;
+}
+
+void RoutingManager::PdPrefixManager::HandleTimer(void)
+{
+    if (mPrefix.IsDeprecated())
+    {
+        WithdrawPrefix();
+    }
+}
+
+Error RoutingManager::PdPrefixManager::Process(const Ip6::Nd::RouterAdvertMessage &aMessage)
+{
+    Error                        error = kErrorNone;
+    DiscoveredPrefixTable::Entry favoredEntry;
+    bool                         currentPrefixUpdated = false;
+
+    VerifyOrExit(aMessage.IsValid(), error = kErrorParse);
+    favoredEntry.Clear();
+
+    for (const Ip6::Nd::Option &option : aMessage)
+    {
+        DiscoveredPrefixTable::Entry entry;
+
+        if (option.GetType() != Ip6::Nd::Option::Type::kTypePrefixInfo ||
+            !static_cast<const Ip6::Nd::PrefixInfoOption &>(option).IsValid())
+        {
+            continue;
+        }
+
+        entry.SetFrom(static_cast<const Ip6::Nd::PrefixInfoOption &>(option));
+
+        // We should accept ULA prefix since it could be used by the internet infrastructure like NAT64.
+        if (entry.GetPrefix().GetLength() > kOmrPrefixLength || entry.GetPrefix().GetLength() == 0 ||
+            entry.GetPrefix().IsLinkLocal() || entry.GetPrefix().IsMulticast())
+        {
+            LogWarn("PdPrefixManager: Ignore invalid PIO entry %s", entry.GetPrefix().ToString().AsCString());
+            continue;
+        }
```

Comment:
Suggestion here: Can add a helper method `PdPrefixManager::IsValidPrefix(const Ip6::Prefix &aPrefix) const` and do all these check in the method? 
It is ok as is but adding such a method can help make the code here easier to read?

Reply:
@erjiaqing , was this addressed?

Reply:
Yes, this is addressed in e605f2224835f8c4c3f4aa44038af2a013cd044e

Reply:
Yes. I think we use the `IsValidPrefix()` now. Looks good to me. Thanks.